### PR TITLE
Add Dockerfile

### PR DESCRIPTION
**Pull request by jfrazelle**
Can be useful for testing everything compiles & building.

**WARNING:**
this does not work yet, but I am no jdbc/flyaway expert so idk

I figured I would open and if you want it you can help me fix, but if not just close this no worries :)

I think the problem is, the model pom.xml is trying to create a database then use it.

Failing in the container with:

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Keywhiz (Parent) ................................... SUCCESS [  7.557 s]
[INFO] Keywhiz Testing .................................... SUCCESS [  8.528 s]
[INFO] Keywhiz API ........................................ SUCCESS [  2.565 s]
[INFO] Keywhiz Client ..................................... SUCCESS [  0.443 s]
[INFO] Keywhiz CLI ........................................ SUCCESS [  6.017 s]
[INFO] Keywhiz HKDF ....................................... SUCCESS [  0.443 s]
[INFO] Keywhiz Model ...................................... FAILURE [  1.460 s]
[INFO] Keywhiz Server ..................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 27.591 s
[INFO] Finished at: 2015-04-28T19:21:04+00:00
[INFO] Final Memory: 41M/707M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.flywaydb:flyway-maven-plugin:3.2:migrate (default) on project keywhiz-model: org.flywaydb.core.api.FlywayException: Unable to obtain Jdbc connection from DataSource (jdbc:postgresql:keywhizdb_test) for user 'root': Connection refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections. -> [Help 1]
[ERROR] 
```

Let me know if you want the full stack trace. But it's probably just some java thing I don't know.

### CODE REVIEW

**Comment by jfrazelle** [comment]
btw, I know it is because postgres is not actually running in the container, but I was wondering if there was a way to skip that test just to build
if not and its super necessary then I don't think this will work this way so feel free to close

**Comment by c0ding** [comment]
Hi there,

@jfrazelle don't you think it'd be best to keep the postgresql part separate (i.e in a second container) and ```--link```the to together?

Cheers.

**Comment by jfrazelle** [comment]
Yes but, I cannot do that on build :(

On Tue, Apr 28, 2015 at 12:44 PM, Martin Simon <notifications@github.com>
wrote:

> Hi there,
>
> @jfrazelle <https://github.com/jfrazelle> don't you think it'd be best to
> keep the postgresql part separate (i.e in a second container) and --linkthe
> to together?
>
> Cheers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/keywhiz/pull/66#issuecomment-97183111>.
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by c0ding** [comment]
True that...
I'd probably go with the ```mvn install``` part in a run.sh file you could call with ```CMD```
If you want to keep everything in one Dockerfile you could add a step (code is untested):

```
RUN \
	service postgresql start && \
	su - postgres -c "psql createdb keywhizdb_development && psql createuser keywhiz"
```

Cheers.

**Comment by jfrazelle** [comment]
there is no way with maven to just ignore the test about creating the db?

On Tue, Apr 28, 2015 at 1:02 PM, Martin Simon <notifications@github.com>
wrote:

> True that...
> I'd probably go with the mvn install part in a run.sh file you could call
> with CMD
> If you want to keep everything in one Dockerfile you could add a step
> (code is untested):
>
> RUN \
>     service postgresql start && \
>     su - postgres -c "psql createdb keywhizdb_development && psql createuser keywhiz"
>
> Cheers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/keywhiz/pull/66#issuecomment-97189589>.
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by c0ding** [comment]
I'm no java expert (quite the opposite actually, total noob inside :smile: ) but I think you can do something like:

```
mvn install -Dmaven.test.skip=true
```

Cheers.

**Comment by alokmenghrajani** [comment]
We can do one of two things: make sure that `-Dmaven.test.skip=true` (or `-DskipTests`) works, or run all the tests using h2 (right now, it's a combination of h2 and postgres). H2 is an embedded SQL engine. I am leaning towards using h2 for all development purpose. It would imply one less thing people have to care about.

**Comment by jfrazelle** [comment]
Ah cool so there is no way to skip just that one test I guess because I
would definitely hate to skip all

On Tuesday, April 28, 2015, Alok Menghrajani <notifications@github.com>
wrote:

> We can do one of two things: make sure that -Dmaven.test.skip=true (or
> -DskipTests) works, or run all the tests using h2 (right now, it's a
> combination of h2 and postgres). H2 is an embedded SQL engine. I am leaning
> towards using h2 for all development purpose. It would imply one less thing
> people have to care about.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/keywhiz/pull/66#issuecomment-97305862>.
>


-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by mbentley** [comment]
@jfrazelle - I have a branch that is able to build successfully by setting up the pg database https://github.com/mbentley/keywhiz/tree/add-dockerfile-hack

It is super hackish right now but I'm hoping to get it cleaned up quite a bit to make it not quite so terrible

**Comment by jfrazelle** [comment]
Does it require running pg in the container w the tests, so it's a
container w multiple processes because you could do that but I might just
turn a blind eye then

On Friday, May 1, 2015, Matt Bentley <notifications@github.com> wrote:

> @jfrazelle <https://github.com/jfrazelle> - I have a branch that is able
> to build successfully by setting up the pg database
> https://github.com/mbentley/keywhiz/tree/add-dockerfile-hack
>
> It is super hackish right now but I'm hoping to get it cleaned up quite a
> bit to make it not quite so terrible
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/keywhiz/pull/66#issuecomment-98194107>.
>


-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by diogomonica** [comment]
@mbentley 

```bash
rm: invalid option -- 'p'
Try 'rm --help' for more information.
INFO[0019] The command [/bin/sh -c apt-get update && apt-get install -y     postgresql     --no-install-recommends     && rm -rf /var/lib/apt/lists/* RUN mkdir -p /usr/src/app] returned a non-zero code: 1
```

**Comment by mbentley** [comment]
@jfrazelle - At build time, yes it runs pg in the background and killing it once it is complete.  It should also probably clean up a bit of junk after it does that but I was hacking on it this morning.  Totally not ideal, hench the `-hack` in the branch name :smile:

@diogomonica - Sorry, I just rebased and squashed my commits so it is fixed now if you pull.  It broke from a missed `\` when I was cleaning things up.

**Comment by diogomonica** [comment]
@sul3n3t @alokmenghrajani thoughts? We would love to have a clean solution for people to setup/run keywhiz server quickly.

Porting the tests to H2 seems like it removes the major hurdle (having to bundle PG in the container).

**Comment by alokmenghrajani** [comment]
Running everything under H2 by default is nice in many ways but I won't be able to get to it right away. I'm ok merging this PR for now if everyone agrees it's a step forward?

**Comment by diogomonica** [comment]
@alokmenghrajani not yet.

@mbentley's Docker file works for building Keywhiz. Right now it doesn't run keywhiz server because it fails to connect to Postgres.

**Comment by mbentley** [comment]
So I did this and I cringed pretty hard but it is something to start from:

Start postgres:
`docker run -d --name postgres -e POSTGRES_USER=root postgres`

Start keywhiz (you need to build this first):
`docker run -it --rm --link postgres:postgres -p 4444:4444 -p 8085:8085 mbentley/keywhiz bash`

Once in the container, run the following:
```
sed -i "s#localhost/keywhizdb_development#${POSTGRES_PORT_5432_TCP_ADDR}/root#g" /usr/src/app/server/src/main/resources/keywhiz-development.yaml
sed -i "s#127.0.0.1/keywhizdb_development#${POSTGRES_PORT_5432_TCP_ADDR}/root#g" /usr/src/app/server/src/main/resources/keywhiz-development.yaml
java -jar server/target/keywhiz-server-*-SNAPSHOT-shaded.jar preview-migrate server/src/main/resources/keywhiz-development.yaml
java -jar server/target/keywhiz-server-*-SNAPSHOT-shaded.jar migrate server/src/main/resources/keywhiz-development.yaml
java -jar server/target/keywhiz-server-*-SNAPSHOT-shaded.jar server server/src/main/resources/keywhiz-development.yaml
```

**Comment by calavera** [comment]
It looks like https://github.com/square/keywhiz/commit/6556985c79175f1f25c1ccbe46a3b9d6b4b26316 broke the docker build because failing to connect to the database is an error in the build process :crying_cat_face:

**Comment by jfrazelle** [comment]
it was never working in the first place ;)

On Wed, May 20, 2015 at 12:04 PM, David Calavera <notifications@github.com>
wrote:

> It looks like 6556985
> <https://github.com/square/keywhiz/commit/6556985c79175f1f25c1ccbe46a3b9d6b4b26316>
> broke the docker build because failing to connect to the database is an
> error in the build process [image: :crying_cat_face:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/keywhiz/pull/66#issuecomment-103997615>.
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by jfrazelle** [comment]
it's against my morals to start a db in the build container....

On Wed, May 20, 2015 at 12:04 PM, Jessica Frazelle <me@jessfraz.com> wrote:

> it was never working in the first place ;)
>
> On Wed, May 20, 2015 at 12:04 PM, David Calavera <notifications@github.com
> > wrote:
>
>> It looks like 6556985
>> <https://github.com/square/keywhiz/commit/6556985c79175f1f25c1ccbe46a3b9d6b4b26316>
>> broke the docker build because failing to connect to the database is an
>> error in the build process [image: :crying_cat_face:]
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/square/keywhiz/pull/66#issuecomment-103997615>.
>>
>
>
>
> --
>
>
> Jessie Frazelle
> 4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
> pgp.mit.edu
> <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by sul3n3t** [comment]
master is a little influx between releases. We're trying to make postgres, mysql, and H2 all work. H2 would be the default test DB. It's embedded and in-memory, so that will remove the concern about starting a DB for build/test.

**Comment by calavera** [comment]
cool, good to know. I don't mind waiting.

**Comment by alokmenghrajani** [comment]
https://github.com/square/keywhiz/pull/92 was merged yesterday and might help you?

**Comment by jfrazelle** [comment]
hell ya, let me update ;)

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+# mkdir for app
+RUN mkdir -p /usr/src/app
+WORKDIR /usr/src/app
+
+# add the source
+ADD . /usr/src/app
+
+# install
+RUN mvn install
```
Comment (alokmenghrajani) [inline Dockerfile]: don't you want `mvn install -P h2`?

Comment (jfrazelle) [inline Dockerfile]: updated my b, i am no maven extraordinaire

On Thu, Jun 4, 2015 at 4:15 PM, Alok Menghrajani
<notifications@github.com> wrote:
> In Dockerfile:
>
>> +#   docker build --rm --force-rm -t square/keywhiz .
>> +#
>> +# Example usage:
>> +#   docker run square/keywhiz java -jar
>> server/target/keywhiz-server-*-SNAPSHOT-shaded.jar server
>> server/src/main/resources/keywhiz-development.yaml
>> +#
>> +FROM maven:3.3-jdk-8
>> +
>> +# mkdir for app
>> +RUN mkdir -p /usr/src/app
>> +WORKDIR /usr/src/app
>> +
>> +# add the source
>> +ADD . /usr/src/app
>> +
>> +# install
>> +RUN mvn install
>
> don't you want mvn install -P h2?
>
> —
> Reply to this email directly or view it on GitHub.



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu

**Comment by alokmenghrajani** [comment]
@diogomonica time to merge this?

**Comment by jfrazelle** [comment]
you can trust me :D

Proof it works: 

```
.... bunch of logs blah.....
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Keywhiz (Parent) ................................... SUCCESS [ 17.628 s]
[INFO] Keywhiz Testing .................................... SUCCESS [ 18.352 s]
[INFO] Keywhiz API ........................................ SUCCESS [  3.973 s]
[INFO] Keywhiz Client ..................................... SUCCESS [  0.738 s]
[INFO] Keywhiz CLI ........................................ SUCCESS [  9.546 s]
[INFO] Keywhiz HKDF ....................................... SUCCESS [  0.644 s]
[INFO] Keywhiz Model ...................................... SUCCESS [  6.315 s]
[INFO] Keywhiz Server ..................................... SUCCESS [01:19 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:17 min
[INFO] Finished at: 2015-06-05T16:27:01+00:00
[INFO] Final Memory: 159M/704M
[INFO] ------------------------------------------------------------------------
 ---> 423a9acee051
Removing intermediate container 76999042827c
Successfully built 423a9acee051

```

**Comment by alokmenghrajani** [comment]
Sounds good, thanks for doing this! I just wanted to make sure we weren't waiting on anything else.

**Pull request merged**